hack/e2e-test/run-e2e-test.sh

```diff
@@ -0,0 +1,461 @@
+      delete_containers $INFRA_PROVIDERS
+      exit 1
+    fi
+    trap container_deletion_trap INT TERM
+
+    echo "Starting e2e tests on k8s cluster"
+
+    failed_providers=""
+    for p in ${INFRA_PROVIDERS//,/ }; do
+      STORAGE_CONTAINER=$TEST_ID PROVIDER=$p ginkgo -v -timeout=30m -mod=vendor test/e2e
+      TEST_RESULT=$?
+      echo "Tests have run for provider $p with result $TEST_RESULT"
+      if [ $TEST_RESULT -ne 0 ] && [ $TEST_RESULT -ne 197 ]; then
+        echo "Tests failed for provider $p"
+        failed_providers="$failed_providers,$p"
+      fi
+    done
+
+    # Remove the containers after the tests
+    delete_containers $INFRA_PROVIDERS
+
+    if [ -n "$failed_providers" ]; then
+      echo "Tests failed for providers: $failed_providers"
+      exit 1
+    else 
+      echo "Tests passed for all providers"
+    fi
+    set -e
+    trap - INT TERM
+  fi
+}
+
+# setup sets up the infrastructure for the providers like building and loading the etcdbr image, creating buckets for the emulators, installing the cli tools to interact with the providers, etc.
+function setup() {
+  if containsElement $STEPS "setup"; then
+    build_and_load
+    # Setup the infrastructure for the providers in parallel
+    pids=()
+    providers=()
+    index=0
+    for p in ${INFRA_PROVIDERS//,/ }; do
+      providers[index]=$p
+      case $p in
+        aws) setup_aws_e2e & pids[index]=$! ;;
+        gcp) setup_gcp_e2e & pids[index]=$! ;;
+        azure) setup_azure_e2e & pids[index]=$! ;;
+        *) echo "Provider: $p is not supported" ;;
+        esac
+        index=$((index+1))
+    done
+    # Wait for all the provider setups to complete
+    for i in ${!pids[@]}; do
+      set +e
+      wait ${pids[$i]}
+      provider_status=$?
+      set -e
+      if [ $provider_status -ne 0 ]; then
+        echo "Setup failed for provider ${providers[$i]}"
+        cleanup_required="true"
+        cleanup
+        exit 1
+      fi
+    done
+    echo "Setup completed for all providers."
+  fi
+}
+
+function create_or_target_cluster() {
+  trap cluster_deletion_trap INT TERM
+  if containsElement $STEPS "setup"; then
+    # Create a kind cluster for the e2e tests
+    make kind-up || exit 1
+    kubectl wait --for=condition=ready node --all
+    if [ ! -r "$KUBECONFIG" ]; then
+      echo "KUBECONFIG is not set or unreadable, please set the KUBECONFIG environment variable to a valid kubeconfig file."
+      cleanup_required="true"
+      cleanup
+      exit 1
+    fi
+  else
+    # Target the existing kind cluster
+    if containsElement $STEPS "test" || containsElement $STEPS "cleanup"; then
+      # Fetch the kubeconfig file for the kind cluster
+      export KUBECONFIG=$(pwd)/hack/e2e-test/infrastructure/kind/kubeconfig
+      if [ ! -r "$KUBECONFIG" ]; then
+        echo "KUBECONFIG is unreadable, please set the KUBECONFIG environment variable to a valid kubeconfig file."
+        exit 1
+      fi
+    fi
+  fi 
+}
```

Comment:
I would not recommend assuming the existence of a running kind cluster, or creating one in its place. The `KUBECONFIG` should be passed by the user, to ensure that these scripts are not run against some production cluster by accident, if that `KUBECONFIG` is set in the developer's terminal environment. If a kind cluster is not already running, it's the caller's responsibility to create that kind cluster and pass its `KUBECONFIG` to `make test-e2e` target